src/puya/ir/_contract_metadata.py

```diff
@@ -227,28 +231,55 @@ def _extract_arc4_methods_and_type_refs(
         if typ is not None
     ]
     methods = dict[awst_nodes.ContractMethod, models.ARC4Method]()
-    for method_name in unique(m.member_name for m in contract.methods):
-        m = contract.resolve_contract_method(method_name)
-        assert m is not None  # shouldn't logically be possible
-        match m.arc4_method_config:
-            case None:
-                pass
-            case awst_nodes.ARC4BareMethodConfig() as bare_method_config:
-                methods[m] = models.ARC4BareMethod(
-                    id=m.full_name, desc=m.documentation.description, config=bare_method_config
-                )
-            case abi_method_config:
-                event_wtypes = event_collector.process_func(m)
-                events = list(map(_wtype_to_struct, event_wtypes))
-                methods[m] = _abi_method_metadata(ctx, contract, m, abi_method_config, events)
-                if m.return_type is not None:
-                    type_refs.append(m.return_type)
-                type_refs.extend(arg.wtype for arg in m.args)
-                for event_struct in event_wtypes:
-                    type_refs.extend(event_struct.types)
+
+    # aggregate errors for the contract so that the user doesn't get incremental errors
+    with _ARC4TypeMapper.collect_errors() as type_mapper:
+        for method_name in unique(m.member_name for m in contract.methods):
+            m = contract.resolve_contract_method(method_name)
+            assert m is not None  # shouldn't logically be possible
+            match m.arc4_method_config:
+                case None:
+                    pass
+                case awst_nodes.ARC4BareMethodConfig() as bare_method_config:
+                    methods[m] = models.ARC4BareMethod(
+                        id=m.full_name, desc=m.documentation.description, config=bare_method_config
+                    )
+                case abi_method_config:
+                    event_wtypes = event_collector.process_func(m)
+                    events = list(map(_wtype_to_struct, event_wtypes))
+                    methods[m] = _abi_method_metadata(
+                        ctx, contract, m, abi_method_config, events, type_mapper
+                    )
+                    if m.return_type is not None:
+                        type_refs.append(m.return_type)
+                    type_refs.extend(arg.wtype for arg in m.args)
+                    for event_struct in event_wtypes:
+                        type_refs.extend(event_struct.types)
     return methods, type_refs
 
 
+@attrs.frozen
+class _ARC4TypeMapper:
+    errors: list[CodeError] = attrs.field(factory=list)
+
+    @classmethod
+    @contextlib.contextmanager
+    def collect_errors(cls) -> Iterator[typing.Self]:
+        collector = cls()
+        yield collector
+        if collector.errors:
```

Comment:
If another `CodeError` is raised by something other than through the `.map(...)` function of this object, then any errors collected will be lost as this statement won't execute.